File: math/leetcode_Divide_Two_Integers.java
/*Divide_Two_Integers
leetcode

Divide two integers without using multiplication, division and mod operator.
*/

public class leetcode_Divide_Two_Integers {

    public static void main(String[] args) {
        System.out.println(divide(2147483647, 1));
    }

    public static int divide(int dividend, int divisor) {
		// Edge cases.
		if (dividend == 0) {
			return 0;
		}
		if (divisor == 1) {
			return dividend;
		}
		
		if (dividend == divisor) {
			return 1;
		}
	
		// Math.abs(Integer.MIN_VALUE) == Integer.MIN_VALUE then the logic below will not work.
		// Two's complement 10000000000000000000000000000000 -> (-1) 011111111111111111111111111111111
		// -> (flip) 10000000000000000000000000000000 becomes itself.
		// Hack the divident by increasing its value to avoid int overflow.	
		if (dividend == Integer.MIN_VALUE) {
			return (divisor > 0) ? divide(dividend + divisor, divisor) - 1 : divide(dividend - divisor, divisor) + 1;
		}
		if (divisor == Integer.MIN_VALUE) {
			return 0;
		}
		boolean neg = (dividend < 0 && divisor > 0) || (dividend > 0 && divisor < 0);
		long dividendLong = Math.abs(dividend);
		long divisorLong = Math.abs(divisor);
	
		// Find the maximum idx that divisor << idx <= dividend
		int idx = 0;
		while ((divisorLong << (idx + 1) ) <= dividendLong) {
			++idx;
		}
		
		int res = 0;
		for (; idx >= 0; --idx) {
			long digitVal = divisorLong << idx;
			// Process as binay digit.
			if (digitVal <= dividendLong) {
				res = res | (1 << idx);
				dividendLong -= digitVal;
			}
		}
		
		return (neg) ? -res : res;
    }
}

File: math/leetcode_Divide_Two_Integers.cpp
/*
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode divide two integers
 */

#include <stdio.h>
#include <stdlib.h>
#include <limits>

using std::numeric_limits;


class Solution {
 public:
  int divide(int dividend, int divisor) {
    // Some edge cases given the nature of dividing.
    if (divisor == 0) {
      throw -1;  // Divide by zero error.
    }

    if (divisor == 1) {
      return dividend;
    }

    if (dividend == divisor) {
      return 1;
    }

    if (dividend == 0) {
      return 0;
    }

    if (dividend == numeric_limits<int>::min() && divisor == -1) {
      return numeric_limits<int>::max();
    }

    // Hanlde int overflow caused by abs.
    if (divisor == numeric_limits<int>::min()) {
      return 0;
    }

    if (dividend == numeric_limits<int>::min()) {
      // using a / b == (a + b) / b - 1 to make sure a + b is not MIN_INT.
      if (divisor > 0) {
        return divide(dividend + divisor, divisor) - 1;
      }
      if (divisor < 0) {
        return divide(dividend - divisor, divisor) + 1;
      }
    }

    bool negative = (dividend > 0 && divisor < 0) || (dividend < 0  && divisor > 0);

    long dividendLong = abs(dividend);
    long divisorLong = abs(divisor);

    int binaryDigits = 0;

    // Find out the max num of digits in binary.
    while (divisorLong << (binaryDigits + 1) <= dividendLong) {
      ++binaryDigits;
    }

    int res = 0;

    // Substracting in term of binary
    while (binaryDigits >= 0) {
      long substractor = divisorLong << binaryDigits;
      if (substractor <= dividendLong) {
        res = res + (1 << binaryDigits);
        dividendLong -= substractor;
      }
      --binaryDigits;
    }

    return negative ? -res : res;
  }
};


int main() {
  Solution sol;
  int res = sol.divide(2147483647, 2);
  printf("Result: %d\n", res);
}
